### PR TITLE
Setup User Table in DB

### PR DESCRIPTION
**Pull request by samau3**
1. Added a database to store user information
2. Updated JWT to use Discord IDs instead of username 
3. Updated routing to handle DB queries and change in JWT information

### CODE REVIEW

File: server/paper-trader-schema.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  balance NUMERIC,
+  is_admin BOOLEAN NOT NULL DEFAULT FALSE
+);
+
+CREATE TABLE transactions (
+  id SERIAL PRIMARY KEY,
+  ticker TEXT NOT NULL,
+  quantity NUMERIC CHECK (quantity >= 0),
+  price NUMERIC CHECK (price >= 0),
+  transction_time TIMESTAMP,
```
Comment (li-andy) [inline server/paper-trader-schema.sql:12]: does this refer to the creation time? If so, this is a bit vaguely named. 

something more standardized like `created_at` would make it more clear that this timestamp isn't being used to differentiate the action being initialized vs the time at which the transaction actually settles

File: server/src/models/user.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import { db } from "../db";
+
+import { BadRequestError } from "../errors";
+
+export class User {
+  // get only the username?
```
Comment (li-andy) [inline server/src/models/user.ts:6]: ?

File: app/src/helpers/PaperTraderApi.ts [hunk 3]
```diff
@@ -29,10 +29,9 @@ class PaperTraderApi {
         }
 
         const result = (await axios.get(
-            `${BACKEND_BASE_URL}/profile`,
+            `${BACKEND_BASE_URL}/profile/username`,
```
Comment (li-andy) [inline app/src/helpers/PaperTraderApi.ts:32]: interesting url

File: server/src/routes/profile.ts [hunk 4]
```diff
@@ -11,12 +12,21 @@ export const router = Router();
...
-router.get('/', ensureCorrectUser, async (req, res, next) => {
+router.get('/username', ensureCorrectUser, async (req, res, next) => {
 	try {
-		// query database for user's data in the future?
-		const user = res.locals.user;
+		const user = await User.login(res.locals.user);
 		return res.json({ user });
 	} catch (err) {
 		return next(err);
 	}
```
Comment (li-andy) [inline server/src/routes/profile.ts:21]: indentation seems massive for this file

File: server/src/routes/profile.ts [hunk 5]
```diff
@@ -11,12 +12,21 @@ export const router = Router();
...
+})
+
+router.get('/summary', ensureCorrectUser, async (req, res, next) => {
+	try {
+		// query database for user's data in the future?
```
Comment (li-andy) [inline server/src/routes/profile.ts:26]: what purpose is this route serving?

File: server/paper-trader-schema.sql (hunk 1 above)
Comment (samau3) [reply server/paper-trader-schema.sql:12]: Ah fair point, will make that change.

File: server/src/routes/profile.ts (hunk 4 above)
Comment (samau3) [reply server/src/routes/profile.ts:21]: Still have no idea why my VSCode formatting is behaving the way it is, tried changing the settings to no avail.

File: server/src/models/user.ts (hunk 2 above)
Comment (samau3) [reply server/src/models/user.ts:6]: Originally I was uncertain if it would be a good idea to just have a request that queries the database only for the user's username, leading to creating a separate database query that just gets the user's balance information. 

Would like some feedback to see if it's still a good idea to have the two queries be separated, but with the current setup this comment isn't needed anymore.

File: app/src/helpers/PaperTraderApi.ts (hunk 3 above)
Comment (li-andy) [reply app/src/helpers/PaperTraderApi.ts:32]: how does this differ from `/profile`?

File: server/src/routes/profile.ts (hunk 5 above)
Comment (li-andy) [reply server/src/routes/profile.ts:26]: might want to use a clearer route name

File: app/src/helpers/PaperTraderApi.ts (hunk 3 above)
Comment (samau3) [reply app/src/helpers/PaperTraderApi.ts:32]: It was just to make differentiate between getting just the username vs the user's account information; 

- `/profile/username` just gets the user's username from the database to store for React.Context
- `/proifle/summary` (will be renamed) will be accessed on the Profile component to get balance, holdings, transactions

However, changing `/profile/username` back to just `/profile` that just returns the username is okay, I was worried there might be issues with confusing pathing - will make the change.

Comment (li-andy) [reply app/src/helpers/PaperTraderApi.ts:32]: it seems unnecessary to have a dedicated route just for fetching the username. Is that something that could be stored client-side instead?

File: server/paper-trader-schema.sql [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  is_admin BOOLEAN NOT NULL DEFAULT FALSE
+);
+
+CREATE TABLE transactions (
+  id SERIAL PRIMARY KEY,
+  ticker TEXT NOT NULL,
+  quantity NUMERIC CHECK (quantity >= 0),
+  price NUMERIC CHECK (price >= 0),
+  created_at TIMESTAMP,
+  username VARCHAR(25) NOT NULL
```
Comment (li-andy) [inline server/paper-trader-schema.sql:13]: Why does this length differ from `username` from the table above?

File: server/paper-trader-schema.sql [hunk 7]
```diff
@@ -0,0 +1,22 @@
+CREATE TABLE users (
+  username VARCHAR(36) PRIMARY KEY,
+  balance NUMERIC,
+  is_admin BOOLEAN NOT NULL DEFAULT FALSE
+);
+
+CREATE TABLE transactions (
+  id SERIAL PRIMARY KEY,
+  ticker TEXT NOT NULL,
```
Comment (li-andy) [inline server/paper-trader-schema.sql:9]: Is there a logical limit for the length of `TEXT`?

Comment (samau3) [reply server/paper-trader-schema.sql:9]: ah will change to `VARCHAR(4)`

File: server/paper-trader-schema.sql (hunk 6 above)
Comment (samau3) [reply server/paper-trader-schema.sql:13]: Definitely a typo, will make change to match

File: server/paper-trader-schema.sql (hunk 7 above)
Comment (li-andy) [reply server/paper-trader-schema.sql]: are there tickers that can be longer than 4 characters?

Comment (samau3) [reply server/paper-trader-schema.sql]: Whoops, originally looked up NYSE tickers and it mentioned they're 4 or less characters (Finnhub uses NYSE) and that NASDAQ would be 5 or less characters. However, one top example of a NTSE stock ticker was BRK.B which is 5 characters for the NYSE.

Updating to account for this.

File: app/src/helpers/PaperTraderApi.ts (hunk 3 above)
Comment (samau3) [reply app/src/helpers/PaperTraderApi.ts:32]: One way would be to decode the JWT that is just the user's username, and store that in React Context? This would remove the need for a route that just returns the user's username.

There will need to be some additional changes to be made in the server routes, such as querying the user's username in the database during the Discord login process to verify they're not a new user, and to add a `/register` route so that they can be added to the user database.

Comment (li-andy) [reply app/src/helpers/PaperTraderApi.ts:32]: are usernames reliable as identifiers? IIUC discord usernames can be changed. If so, user IDs would be what you should key off of

Comment (samau3) [reply app/src/helpers/PaperTraderApi.ts:32]: Oh yes Discord usernames can be changed, will need to change to utilize their user ID.

Comment (samau3) [reply app/src/helpers/PaperTraderApi.ts:32]: Updated authentication structure so there isn't any unnecessary JWT verification and sending back and forth

File: server/src/models/user.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        VALUES ($1, $2, $3, $4)
+        RETURNING id`,
+      [id, username, balance, is_admin]
+    )
+
+    const user = result.rows[0];
+    return user;
+  }
+
+  static async getProfile(id: string): Promise<string> {
```
Comment (samau3) [inline server/src/models/user.ts:39]: There is some inconsistency between profile/portfolio usage, will need to work on using one or having more defined distinctions between the two.

File: server/paper-trader-schema.sql [hunk 9]
```diff
@@ -0,0 +1,23 @@
+CREATE TABLE users (
+  id VARCHAR PRIMARY KEY,
+  username VARCHAR(36),
```
Comment (li-andy) [inline server/paper-trader-schema.sql:3]: will this mirror a user's discord username? if so, will the discriminator be included?

File: server/src/models/user.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
+import { db } from "../db";
+
+export class User {
+  static async login(id: string, username: string): Promise<string> {
```
Comment (li-andy) [inline server/src/models/user.ts:4]:
```suggestion
  static async loginOrRegister(id: string, username: string): Promise<string> {
```
this naming might be more reflective of what's happening in the function

File: server/src/models/user.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  static async getProfile(id: string): Promise<string> {
+    const result = await db.query(
+      `SELECT username, balance
+        FROM users
+        WHERE id=$1`,
+      [id]
+    );
+    const balance = result.rows[0];
```
Comment (li-andy) [inline server/src/models/user.ts:46]: is balance guaranteed to exist?

File: server/src/routes/login.ts [hunk 12]
```diff
@@ -35,9 +36,11 @@ router.get('/', async (req, res, next) => {
 	try {
 		const userResult = await Discord.getDiscordUser(oauthTokenData.token_type, oauthTokenData.access_token);
 
-		const { username, discriminator } = userResult;
+		const { id, username, discriminator } = userResult;
 
-		const token = createToken(`${username}${discriminator}`);
+		const user = await User.login(id, `${username}${discriminator}`);
```
Comment (li-andy) [inline server/src/routes/login.ts:41]: does `discriminator` already include the `#`?

File: server/src/routes/stock.ts [hunk 13]
```diff
@@ -11,9 +10,8 @@ export const router = Router();
  * Returns stock quote information if a valid ticker is provided.
  * Otherwise, returns BadRequestError.
  *
- * Authorization required: ensureCorrectUser
  */
-router.get('/search', ensureCorrectUser, async (req, res, next) => {
+router.get('/search', async (req, res, next) => {
```
Comment (li-andy) [inline server/src/routes/stock.ts:14]: do these need to be replaced with a different middleware?

File: server/paper-trader-schema.sql (hunk 9 above)
Comment (samau3) [reply server/paper-trader-schema.sql:3]: It will, but I have been excluding the "#" from the count; will update to include

File: server/src/models/user.ts (hunk 11 above)
Comment (samau3) [reply server/src/models/user.ts:46]: At the moment, when users register they start with a default of 10,000 for balance. Will need to update the table schema for a default value of zero though, in case there are plans in the future to start users with a zero balance instead.

File: server/src/routes/login.ts (hunk 12 above)
Comment (samau3) [reply server/src/routes/login.ts:41]: Does not, will need to include

File: server/src/routes/stock.ts (hunk 13 above)
Comment (samau3) [reply server/src/routes/stock.ts:14]: I'll need to make further changes to the existing authenticateJWT middleware as right not it's not properly protecting all routes that need to be protected.

File: server/src/models/user.ts (hunk 10 above)
Comment (samau3) [reply server/src/models/user.ts:4]: Good point, will make the change!

File: server/src/routes/stock.ts (hunk 13 above)
Comment (samau3) [reply server/src/routes/stock.ts:14]: Updated authenticateJWT to properly replace ensureCorrectUser without impacting login requests that don't have a JWT yet.

File: server/paper-trader-schema.sql [hunk 14]
```diff
@@ -0,0 +1,23 @@
+CREATE TABLE users (
+  id VARCHAR PRIMARY KEY,
+  username VARCHAR(37),
+  balance NUMERIC DEFAULT 0,
+  is_admin BOOLEAN NOT NULL DEFAULT FALSE
```
Comment (li-andy) [inline server/paper-trader-schema.sql:5]: what do you plan on doing with this flag?

File: server/src/models/user.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        WHERE id=$1`,
+      [id]
+    );
+
+    const user = result.rows[0];
+
+    if (user) {
+      return user;
+    } else {
+      this.register(id, username, 10000, false);
```
Comment (li-andy) [inline server/src/models/user.ts:18]: maybe pull out the `10000` from being a hardcoded value into a config value?

File: server/paper-trader-schema.sql (hunk 14 above)
Comment (samau3) [reply server/paper-trader-schema.sql:5]: I was thinking of having admin accounts be able to manually adjust a user's balance. Might remove in the future if there's not much benefit for being an admin though.

File: server/src/routes/profile.ts (hunk 4 above)
Comment (samau3) [reply server/src/routes/profile.ts]: Figured it out! Had to change settings from using Tabs to Spaces. In VSCode `Tabs: 2` look the same as `Spaces: 2` so needed to apply the `Spaces` formatting to files.

File: server/src/errors.ts [hunk 16]
```diff
@@ -1,17 +1,17 @@
...
+  Accept: string,
+  authorization: string
 }
 
 interface request {
-    _header: string
+  _header: string
 }
 
 /** ExpressError extends normal JS error so we can
```
Comment (li-andy) [inline server/src/errors.ts:17]: next time do this in a separate PR

**Review by li-andy — APPROVED** [review]
left a comment about formatting changes
